service/src/main/java/bio/terra/workspace/service/workspace/flight/DeleteControlledAzureResourcesStep.java

```diff
@@ -0,0 +1,67 @@
+package bio.terra.workspace.service.workspace.flight;
+
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.db.ResourceDao;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.resource.controlled.ControlledResourceService;
+import bio.terra.workspace.service.resource.controlled.model.ControlledResource;
+import bio.terra.workspace.service.workspace.model.CloudPlatform;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * A step to delete all controlled Azure resources resources in a workspace. This reads the list of
+ * controlled Azure resources in a workspace from the WSM database.
+ */
+public class DeleteControlledAzureResourcesStep implements Step {
+
+  private final Logger logger = LoggerFactory.getLogger(DeleteControlledAzureResourcesStep.class);
+  private final ResourceDao resourceDao;
+  private final ControlledResourceService controlledResourceService;
+  private final UUID workspaceUuid;
+  private final AuthenticatedUserRequest userRequest;
+
+  public DeleteControlledAzureResourcesStep(
+      ResourceDao resourceDao,
+      ControlledResourceService controlledResourceService,
+      UUID workspaceUuid,
+      AuthenticatedUserRequest userRequest) {
+    this.resourceDao = resourceDao;
+    this.controlledResourceService = controlledResourceService;
+    this.workspaceUuid = workspaceUuid;
+    this.userRequest = userRequest;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+
+    List<ControlledResource> controlledResourceList =
+        resourceDao.listControlledResources(workspaceUuid, CloudPlatform.AZURE);
+
+    /**
+     * TODO: https://broadworkbench.atlassian.net/browse/WOR-92 delete Azure storage containers
+     * first before deleting storage account to ensure Sam and WSM DBs are cleaned up
+     */
+    for (ControlledResource resource : controlledResourceList) {
+      controlledResourceService.deleteControlledResourceSync(
```

Comment:
This will throw an exception if the delete fails. That will cause this flight to fail. Is that the behavior you want?
That failure will not be retried. Stairway will undo and fail the flight.

An alternative would be to catch and track the failures, but continue deleting as many resources as you can. If the failures are retry-able, then you could return Retry. That would maybe succeed in the face of transient errors.

Reply:
We opted to remove the retry rule here so that if deleting a resource fails the flight will fail. The exception handling when deleting Azure resources just retries all failures at the moment. We are hesitant to do too much retrying here as a result since we expect that simple errors like a resource not existing will lead to it retrying unnecessarily. We've filed [this ticket](https://broadworkbench.atlassian.net/browse/WOR-327) to improve the Azure resource deletion error handling, but we believe this is sufficient for now